HYDRA-1328 : Make maya hydra compilable with maya 2024 devkit (#237)
Autodesk/maya-hydra: doc/rebuildingWithLatestOpenUSDAndPreviousMaya.md

# Building with the latest version of OpenUSD and with a previous version of Maya such as 2024

> Let's take as an example, you want to rebuild MayaHydra in <b>Maya 2024</b> and <b>OpenUSD 24.11</b> under Windows.<br>

Reviewer: ```suggestion
Let's take as an example rebuilding MayaHydra in <b>Maya 2024</b> and <b>OpenUSD 24.11</b> under Windows.<br>
```

With Maya 2024 the OpenUSD version used by default was 22.11 which was compiled with Python 3.10.8.

## Rebuilding OpenUSD

If OpenUSD was not built with the same version of Python as the one used in your Maya version, you need to rebuild it with the suitable Python :<br>

|        Maya version       | Python version   |
|:-------------------------:|:----------------:|
| 2024                 | 3.10.8           |
| 2025                 | 3.11.4           |
| PR                   | 3.11.9           |

So for Maya 2024, we need to rebuild OpenUSD with Python 3.10.8.<br>
To do so, you need to clone the OpenUSD repository and checkout the version used by MayaUsd and MayaHydra which is [v24.11-MayaUsd-Public](https://github.com/autodesk-forks/USD/tree/v24.11-MayaUsd-Public).<br>

> As an example is given a Windows batch file to rebuild OpenUSD with Python 3.10.8 which is provided by mayapy.exe in the Maya/bin folder:<br><br>

Reviewer: ```suggestion
As an example here is a Windows batch file to rebuild OpenUSD with Python 3.10.8 which is provided by mayapy.exe in the Maya/bin folder:<br><br>
```

> But first, you may need to install Python modules for this version of Python if it's not yet installed (this, not only to rebuild OpenUSD but MayaUsd also), they are : <br>

Reviewer: ```suggestion
First, to rebuild both OpenUSD and MayaUsd, you may need to install the following Python modules for this version of Python: <br>
```

PyOpenGL==3.1.0, PySide6, jinja2 and pybind11.<br>
We are assuming that Maya 2024 is installed in its default folder such as C:\Program Files\Autodesk\Maya2024.<br>
```
"C:\Program Files\Autodesk\Maya2024\bin\mayapy.exe" -m pip install PyOpenGL==3.1.0
"C:\Program Files\Autodesk\Maya2024\bin\mayapy.exe" -m pip install PySide6
"C:\Program Files\Autodesk\Maya2024\bin\mayapy.exe" -m pip install jinja2
"C:\Program Files\Autodesk\Maya2024\bin\mayapy.exe" -m pip install pybind11
```
To rebuild OpenUSD see the following Windows batch file as an example :
<br>
```
rem rebuild OpenUSD with Python 3.10.8 for maya 2024

> rem I had to install in mayapy.exe the modules : PyOpenGL==3.1.0, PySide6 and pybind11.

Reviewer: ```suggestion
rem The following modules may need to be installed in mayapy.exe: PyOpenGL==3.1.0, PySide6 and pybind11.
```

set openusd_maya_location=C:/Program^ Files/Autodesk/Maya2024
set openusd_python_include=%openusd_maya_location%/include/Python310/Python
set openusd_python_exe=%openusd_maya_location%/bin/mayapy.exe
set openusd_python_lib=%openusd_maya_location%/lib/python310.lib

rem our destination folder for the build
set destination_folder="D:/USD-24.11-Python3.10.8"

> rem rebuylding OpenUSD with Python 3.10.8 in release with debug info configuration, if you want to use release only, please change the build-variant

Reviewer: ```suggestion
rem rebuilding OpenUSD with Python 3.10.8 in release with debug info configuration, if you want to use release only, please change the build-variant
```

```
rem Save the current PATH as we are going to modify it
set original_path=%PATH%

rem OpenUSD 24.11 rebuilt with Python 3.10.8
set mayausd_pixar_usd_location=D:/USD-24.11-Python3.10.8

rem download and install the Maya 2024 devkit in a folder
set mayausd_maya_devkit_location=D:/Autodesk_Maya_2024_DEVKIT_Windows/devkitBase

rem Maya 2024 location
set mayausd_maya_location=C:/Program^ Files/Autodesk/Maya2024
set mayausd_python_include=%mayausd_maya_location%/include/Python310/Python
set mayausd_python_exe=%mayausd_maya_location%/bin/mayapy.exe
set mayausd_python_lib=%mayausd_maya_location%/lib/python310.lib

rem Remove other Python installations from the PATH to avoid conflict 
rem As I had Python 3.11 installed. It would also be possible to use a Python virtual environment: https://docs.python.org/3/library/venv.html
set PATH=C:\Program Files\Autodesk\Maya2024\bin;%original_path%
set PATH=%PATH:C:\Users\%username%\AppData\Local\Programs\Python\Python311\Scripts;=%
set PATH=%PATH:C:\Users\%username%\AppData\Local\Programs\Python\Python311;=%
set PATH=%PATH:C:\Users\%username%\AppData\Local\Programs\Python\Python311\=%

> rem Add OpenUSD bon and lib path to the Windows PATH

Reviewer: ```suggestion
rem Add OpenUSD bin and lib path to the Windows PATH
```

set mayausd_qtversion=6.5.3

rem Output diagnostic information
echo PATH=%PATH%
echo PYTHONPATH=%PYTHONPATH%

rem Use the mayapy.exe directly for the build process
"%mayausd_python_exe%" build.py "%mayausd_destination%_" ^
	-v=3 ^
	--stages=clean,configure,build,install ^
	--pxrusd-location=%mayausd_pixar_usd_location% ^
	--build-relwithdebug  ^
	--maya-location="%mayausd_maya_location%" ^
	--generator="Visual Studio 17 2022" ^
	--devkit-location="%mayausd_maya_devkit_location%" ^
	--materialx ^
	--build-args="-DPYTHON_INCLUDE_DIR=\"%mayausd_python_include%\",-DPython_EXECUTABLE=\"%mayausd_python_exe%\",-DPython3_VERSION=3.10.8,-DPYTHON_LIBRARIES=\"%mayausd_python_lib%\",-DBUILD_WITH_PYTHON_3=ON,-DBUILD_WITH_PYTHON_3_VERSION=3.10,-DBUILD_TESTS=ON,-DBUILD_PXR_PLUGIN=OFF,-DBUILD_AL_PLUGIN=OFF,-DQT_LOCATION=\"%mayausd_qtlocation%\",-DQT_VERSION=%mayausd_qtversion%,-DPython3_ROOT_DIR=\"%mayausd_maya_location%\",-DPython3_FIND_STRATEGY=LOCATION,-DPython3_INCLUDE_DIR=\"%mayausd_python_include%\",-DPython3_LIBRARY=\"%mayausd_python_lib%\""

rem Restore the original PATH
set PATH=%original_path%

rem Unset the Python environment variables
set PYTHONPATH=
```

Note : <br>

> In my case, I had an issue when rebuilding MayaUsd which was : https://github.com/PixarAnimationStudios/OpenUSD/issues/3310 <br>

Reviewer: ```suggestion
You may have the following issue when rebuilding MayaUsd: https://github.com/PixarAnimationStudios/OpenUSD/issues/3310 <br>
```

> So I removed "OpenGL::GL" from the cmake file :<br>

Reviewer: ```suggestion
In this case remove "OpenGL::GL" from the cmake file :<br>
```

D:/USD-24.11-Python3.10.8/cmake/pxrTargets.cmake <br>

> And when rebuilding MayaUsd, I had to manually add the link to "OpenGL32.lib" to the MayaUsd project under Visual Studio in order to link correctly.

Reviewer: ```suggestion
When rebuilding MayaUsd, manually add the link to "OpenGL32.lib" to the MayaUsd project under Visual Studio in order to link correctly.
```

## Rebuilding MayaHydra

> Now you have OpenUSD and MayaUsd rebuilt with the same version of Python as Maya 2024, you can rebuild MayaHydra.<br>

Reviewer: ```suggestion
OpenUSD and MayaUsd are now rebuilt with the same version of Python as Maya 2024.  MayaHydra can now be built.<br>
```

> # Building with the latest version of OpenUSD and with a previous version of Maya such as 2024

Author: This file can bee seen as (for better readibility) :
https://git.autodesk.com/maya3d/maya-hydra/blob/lanierd/HYDRA-1328/doc/rebuildingWithLatestOpenUSDAndPreviousMaya.md
